bindings/matrix-sdk-ffi/src/client_builder.rs

```diff
@@ -27,6 +33,140 @@ enum HomeserverConfig {
     ServerNameOrUrl(String),
 }
 
+/// Data for the QR code login mechanism.
+///
+/// The [`QrCodeData`] can be serialized and encoded as a QR code or it can be
+/// decoded from a QR code.
+#[derive(Debug, uniffi::Object)]
+pub struct QrCodeData {
+    inner: qrcode::QrCodeData,
+}
+
+#[uniffi::export]
+impl QrCodeData {
+    /// Attempt to decode a slice of bytes into a [`QrCodeData`] object.
+    ///
+    /// The slice of bytes would generally be returned by a QR code decoder.
+    #[uniffi::constructor]
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Arc<Self>, QrCodeDecodeError> {
+        Ok(Self { inner: qrcode::QrCodeData::from_bytes(&bytes)? }.into())
+    }
+}
+
+/// Error type for the decoding of the [`QrCodeData`].
+#[derive(Debug, thiserror::Error, uniffi::Error)]
+#[uniffi(flat_error)]
+pub enum QrCodeDecodeError {
+    #[error("Error decoding QR code: {error:?}")]
+    Crypto {
+        #[from]
+        error: LoginQrCodeDecodeError,
+    },
+}
+
+#[derive(Debug, thiserror::Error, uniffi::Error)]
+pub enum HumanQrLoginError {
```

Comment:
Is the intent to display those errors to the end-user? If so, I think it's not our task, since these error messages have to be localized, and it's likely we're not the best equipped for crafting the most understandable error messages ("what's a sliding sync? what's a homeserver?"), although I appreciate the intent to have precise errors that are nicely documented :+1: 

(No special action required; maybe "downgrade" the `error()` to doc comments, but maybe the apps may use these error messages directly, not sure.)

Reply:
We used to have two error types, the full Rust one and this one, and a method to convert from one to the other, but uniffi wouldn't comply as of now with that strategy. We might want to try to bring the full error type back once we add the other QR login flows.

The applications will not use the error messages, though each of those variants will create a different error screen.

That being said, I can imagine some other application which would not care about localizations to use the error message directly.